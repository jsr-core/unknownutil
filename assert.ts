import type { Predicate } from "./type.ts";

/**
 * A factory function that generates assertion error messages.
 */
export type AssertMessageFactory = (
  x: unknown,
  pred: Predicate<unknown>,
  name?: string,
) => string;

/**
 * The default factory function used to generate assertion error messages.
 */
export const defaultAssertMessageFactory: AssertMessageFactory = (
  x,
  pred,
  name,
) => {
  const p = pred.name || "anonymous predicate";
  const t = typeof x;
  const v = JSON.stringify(x, null, 2);
  return `Expected ${
    name ?? "a value"
  } that satisfies the predicate ${p}, got ${t}: ${v}`;
};

let assertMessageFactory = defaultAssertMessageFactory;

/**
 * Represents an error that occurs when an assertion fails.
 */
export class AssertError extends Error {
  /**
   * Constructs a new instance.
   * @param message The error message.
   */
  constructor(message?: string) {
    super(message);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AssertError);
    }

    this.name = this.constructor.name;
  }
}

/**
 * Sets the factory function used to generate assertion error messages.
 *
 * ```ts
 * import { is, setAssertMessageFactory } from "@core/unknownutil";
 *
 * setAssertMessageFactory((x, pred) => {
 *   if (pred === is.String) {
 *     return `Expected a string, got ${typeof x}`;
 *   } else if (pred === is.Number) {
 *     return `Expected a number, got ${typeof x}`;
 *   } else if (pred === is.Boolean) {
 *     return `Expected a boolean, got ${typeof x}`;
 *   } else {
 *     return `Expected a value that satisfies the predicate, got ${typeof x}`;
 *   }
 * });
 * ```
 *
 * @param factory The factory function.
 */
export function setAssertMessageFactory(factory: AssertMessageFactory): void {
  assertMessageFactory = factory;
}

/**
 * Asserts that the given value satisfies the provided predicate.
 *
 * It throws {@linkcode AssertError} if the value does not satisfy the predicate.
 *
 * ```ts
 * import { assert, is } from "@core/unknownutil";
 *
 * const a: unknown = "hello";
 * assert(a, is.String);
 * // a is now narrowed to string
 * ```
 *
 * @param x The value to be asserted.
 * @param pred The predicate function to test the value against.
 * @param options Optional configuration for the assertion.
 * @returns The function has a return type of `asserts x is T` to help TypeScript narrow down the type of `x` after the assertion.
 */
export function assert<T>(
  x: unknown,
  pred: Predicate<T>,
  options: { message?: string; name?: string } = {},
): asserts x is T {
  if (!pred(x)) {
    throw new AssertError(
      options.message ?? assertMessageFactory(x, pred, options.name),
    );
  }
}
