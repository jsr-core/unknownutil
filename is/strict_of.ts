import { rewriteName } from "../_funcutil.ts";
import type { WithPredObj } from "../_annotation.ts";
import type { Predicate } from "../type.ts";

/**
 * Return a type predicate function that returns `true` if the type of `x` is strictly follow the `ObjectOf<T>`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.StrictOf(is.ObjectOf({
 *   a: is.Number,
 *   b: is.String,
 *   c: as.Optional(is.Boolean),
 * }));
 * const a: unknown = { a: 0, b: "a", other: "other" };
 * if (isMyType(a)) {
 *   // This block will not be executed because of "other" key in `a`.
 * }
 * ```
 */
export function isStrictOf<
  T extends Record<PropertyKey, unknown>,
  P extends Record<PropertyKey, Predicate<unknown>>,
>(
  pred:
    & Predicate<T>
    & WithPredObj<P>,
):
  & Predicate<T>
  & WithPredObj<P> {
  const s = new Set(Object.keys(pred.predObj));
  return rewriteName(
    (x: unknown): x is T => {
      if (!pred(x)) return false;
      // deno-lint-ignore no-explicit-any
      const ks = Object.keys(x as any);
      return ks.length <= s.size && ks.every((k) => s.has(k));
    },
    "isStrictOf",
    pred,
  ) as Predicate<T> & WithPredObj<P>;
}
