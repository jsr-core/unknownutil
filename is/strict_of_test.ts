import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType, type IsExact } from "@std/testing/types";
import { as } from "../as/mod.ts";
import { is } from "./mod.ts";
import { isStrictOf } from "./strict_of.ts";

Deno.test("isStrictOf<T>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(
      t,
      isStrictOf(is.ObjectOf({ a: is.Number, b: is.String, c: is.Boolean }))
        .name,
    );
    await assertSnapshot(
      t,
      isStrictOf(is.ObjectOf({ a: (_x): _x is string => false })).name,
    );
    await assertSnapshot(
      t,
      isStrictOf(
        is.ObjectOf({
          a: isStrictOf(
            is.ObjectOf({ b: isStrictOf(is.ObjectOf({ c: is.Boolean })) }),
          ),
        }),
      ).name,
    );
  });

  await t.step("returns true on T object", () => {
    const predObj = {
      a: is.Number,
      b: is.String,
      c: is.Boolean,
    };
    assertEquals(
      isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a", c: true }),
      true,
    );
  });

  await t.step("returns false on non T object", () => {
    const predObj = {
      a: is.Number,
      b: is.String,
      c: is.Boolean,
    };
    assertEquals(
      isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a", c: "" }),
      false,
      "Object have a different type property",
    );
    assertEquals(
      isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a" }),
      false,
      "Object does not have one property",
    );
    assertEquals(
      isStrictOf(is.ObjectOf(predObj))({
        a: 0,
        b: "a",
        c: true,
        d: "invalid",
      }),
      false,
      "Object have an unknown property",
    );
  });

  await t.step("predicated type is correct", () => {
    const predObj = {
      a: is.Number,
      b: is.String,
      c: is.Boolean,
    };
    const a: unknown = { a: 0, b: "a", c: true };
    if (isStrictOf(is.ObjectOf(predObj))(a)) {
      assertType<IsExact<typeof a, { a: number; b: string; c: boolean }>>(true);
    }
  });

  await t.step("with optional properties", async (t) => {
    await t.step("returns true on T object", () => {
      const predObj = {
        a: is.Number,
        b: is.UnionOf([is.String, is.Undefined]),
        c: as.Optional(is.Boolean),
      };
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a", c: true }),
        true,
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a" }),
        true,
        "Object does not have an optional property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a", c: undefined }),
        true,
        "Object has `undefined` as value of optional property",
      );
    });

    await t.step("returns false on non T object", () => {
      const predObj = {
        a: is.Number,
        b: is.UnionOf([is.String, is.Undefined]),
        c: as.Optional(is.Boolean),
      };
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a", c: "" }),
        false,
        "Object have a different type property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, b: "a", c: null }),
        false,
        "Object has `null` as value of optional property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({
          a: 0,
          b: "a",
          c: true,
          d: "invalid",
        }),
        false,
        "Object have an unknown property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({
          a: 0,
          b: "a",
          d: "invalid",
        }),
        false,
        "Object have the same number of properties but an unknown property exists",
      );
    });

    await t.step("predicated type is correct", () => {
      const predObj = {
        a: is.Number,
        b: is.UnionOf([is.String, is.Undefined]),
        c: as.Optional(is.Boolean),
      };
      const a: unknown = { a: 0, b: "a" };
      if (isStrictOf(is.ObjectOf(predObj))(a)) {
        assertType<
          IsExact<
            typeof a,
            { a: number; b: string | undefined; c?: boolean | undefined }
          >
        >(true);
      }
    });
  });

  await t.step("with symbol properties", async (t) => {
    const b = Symbol("b");
    const c = Symbol("c");
    const predObj = {
      a: is.Number,
      [b]: is.UnionOf([is.String, is.Undefined]),
      [c]: as.Optional(is.Boolean),
    };
    await t.step("returns properly named predicate function", async (t) => {
      await assertSnapshot(
        t,
        isStrictOf(
          is.ObjectOf({ a: is.Number, [b]: is.String, [c]: is.Boolean }),
        )
          .name,
      );
      await assertSnapshot(
        t,
        isStrictOf(
          is.ObjectOf({
            a: isStrictOf(
              is.ObjectOf({
                [b]: isStrictOf(is.ObjectOf({ [c]: is.Boolean })),
              }),
            ),
          }),
        ).name,
      );
    });

    await t.step("returns true on T object", () => {
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, [b]: "a", [c]: true }),
        true,
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, [b]: "a" }),
        true,
        "Object does not have an optional property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, [b]: "a", [c]: undefined }),
        true,
        "Object has `undefined` as value of optional property",
      );
    });

    await t.step("returns false on non T object", () => {
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, [b]: "a", [c]: "" }),
        false,
        "Object have a different type property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({ a: 0, [b]: "a", [c]: null }),
        false,
        "Object has `null` as value of optional property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({
          a: 0,
          [b]: "a",
          [c]: true,
          d: "invalid",
        }),
        false,
        "Object have an unknown property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({
          a: 0,
          [b]: "a",
          [c]: true,
          [Symbol("d")]: "invalid",
        }),
        false,
        "Object have an unknown symbol property",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({
          a: 0,
          [b]: "a",
          d: "invalid",
        }),
        false,
        "Object have the same number of properties but an unknown property exists",
      );
      assertEquals(
        isStrictOf(is.ObjectOf(predObj))({
          a: 0,
          [b]: "a",
          [Symbol("d")]: "invalid",
        }),
        false,
        "Object have the same number of properties but an unknown symbol property exists",
      );
    });

    await t.step("predicated type is correct", () => {
      const a: unknown = { a: 0, [b]: "a" };
      if (isStrictOf(is.ObjectOf(predObj))(a)) {
        assertType<
          IsExact<
            typeof a,
            { a: number; [b]: string | undefined; [c]?: boolean | undefined }
          >
        >(true);
      }
    });
  });
});
