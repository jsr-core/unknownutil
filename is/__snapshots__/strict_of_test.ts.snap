export const snapshot = {};

snapshot[`isStrictOf<T> > returns properly named predicate function 1`] = `
"isStrictOf(isObjectOf({
  a: isNumber,
  b: isString,
  c: isBoolean
}))"
`;

snapshot[`isStrictOf<T> > returns properly named predicate function 2`] = `"isStrictOf(isObjectOf({a: a}))"`;

snapshot[`isStrictOf<T> > returns properly named predicate function 3`] = `
"isStrictOf(isObjectOf({
  a: isStrictOf(isObjectOf({
    b: isStrictOf(isObjectOf({c: isBoolean}))
  }))
}))"
`;

snapshot[`isStrictOf<T> > with symbol properties > returns properly named predicate function 1`] = `
"isStrictOf(isObjectOf({
  a: isNumber,
  Symbol(b): isString,
  Symbol(c): isBoolean
}))"
`;

snapshot[`isStrictOf<T> > with symbol properties > returns properly named predicate function 2`] = `
"isStrictOf(isObjectOf({
  a: isStrictOf(isObjectOf({
    Symbol(b): isStrictOf(isObjectOf({Symbol(c): isBoolean}))
  }))
}))"
`;
